packages/core/src/actions.ts

```diff
@@ -582,107 +582,142 @@ export function resolveActions<TContext, TEvent extends EventObject>(
+
+            if (!IS_PRODUCTION) {
+              // warn after resolving as we can create better contextual message here
+              warn(
+                !isString(actionObject.delay) ||
+                  typeof sendAction.delay === 'number',
+                // tslint:disable-next-line:max-line-length
+                `No delay reference for delay expression '${actionObject.delay}' was found on machine '${machine.id}'`
+              );
+            }
+
+            return sendAction;
+          case actionTypes.log:
+            return resolveLog(
+              actionObject as LogAction<TContext, TEvent>,
+              updatedContext,
+              _event
+            );
+          case actionTypes.choose: {
+            const chooseAction = actionObject as ChooseAction<TContext, TEvent>;
+            const matchedActions = chooseAction.conds.find((condition) => {
+              const guard = toGuard(condition.cond, machine.options.guards);
+              return (
+                !guard ||
+                evaluateGuard(
+                  machine,
+                  guard,
+                  updatedContext,
+                  _event,
+                  currentState as any
+                )
+              );
+            })?.actions;
+
+            if (!matchedActions) {
+              return [];
+            }
+
+            const [
+              resolvedActionsFromChoose,
+              resolvedContextFromChoose
+            ] = resolveActions(
+              machine,
+              currentState,
+              updatedContext,
+              _event,
+              toActionObjects(toArray(matchedActions), machine.options.actions)
             );
+            updatedContext = resolvedContextFromChoose;
+            preservedContexts?.push(updatedContext);
+            return resolvedActionsFromChoose;
           }
-
-          return sendAction;
-        case actionTypes.log:
-          return resolveLog(
-            actionObject as LogAction<TContext, TEvent>,
-            updatedContext,
-            _event
-          );
-        case actionTypes.choose: {
-          const chooseAction = actionObject as ChooseAction<TContext, TEvent>;
-          const matchedActions = chooseAction.conds.find((condition) => {
-            const guard = toGuard(condition.cond, machine.options.guards);
-            return (
-              !guard ||
-              evaluateGuard(
-                machine,
-                guard,
-                updatedContext,
-                _event,
-                currentState as any
-              )
+          case actionTypes.pure: {
+            const matchedActions = (actionObject as PureAction<
+              TContext,
+              TEvent
+            >).get(updatedContext, _event.data);
+            if (!matchedActions) {
+              return [];
+            }
+            const [resolvedActionsFromPure, resolvedContext] = resolveActions(
```

Comment:
`preserveActionOrder` should be passed to this `resolveActions` call as well and the appropriate test for this should be added

Reply:
5ae1959d41ac90eab15a46c57e135d59dd2ca05f